### PR TITLE
Support GANG and FCFS distributed mode

### PR DESCRIPTION
**Pull request by zuston**
Linked [issue](https://github.com/linkedin/TonY/issues/501)

### CODE REVIEW

**Comment by zuston** [comment]
please review @oliverhu @burgerkingeater @hungj

**Comment by oliverhu** [comment]
@zuston could you resolve the merge conflicts here?

File: tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java [hunk 1]
```diff
@@ -853,38 +859,53 @@ public String registerWorkerSpec(String taskId, String spec) throws IOException
...
-              LOG.error(errorMsg);
-              session.setFinalStatus(FinalApplicationStatus.FAILED, errorMsg);
-              stop();
-            } else {
-                LOG.info(String.format("Awaiting registration from task %s %s in %s on host %s",
+      // two distributed mode(default is cluster) cases:
+      // 1. when in independent-task mode, task will be allowed to run when AM accept worker registered spec,
+      // 2. when in cluster mode, it will start until all tasks have registered.
+      switch (distributedMode) {
+        case CLUSTER:
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java:866]: CLUSTER, INDEPENDENT_TASK is not accurate, cluster mode doesn't have to wait either. What about GANG and FCFS modes?

File: tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java [hunk 2]
```diff
@@ -390,4 +404,11 @@ private void skewAndHangIfTesting() {
       }
     }
   }
+
+  private boolean isClusterMode() {
+    if (distributedMode == TonyConfigurationKeys.DistributedMode.CLUSTER) {
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java:409]: minor: `return distributedMode == TonyConfigurationKeys.DistributedMode.CLUSTER `

File: tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java (hunk 1 above)
Comment (zuston) [reply tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java:866]: Good. FCFS means first come first schedule?

File: tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java (hunk 2 above)
Comment (zuston) [reply tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java]: Done

File: tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java (hunk 1 above)
Comment (zuston) [reply tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java]: Done

File: tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java [hunk 3]
```diff
@@ -173,15 +173,23 @@ private static TaskExecutor createExecutor() throws Exception {
       LOG.error("Failed to register worker with AM.");
       throw new Exception("Failed to register worker with AM.");
     }
+    LOG.info("Task is on distributed mode: " + executor.distributedMode);
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java:176]: log.debug

**Review by oliverhu — APPROVED** [review]

File: tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java (hunk 3 above)
Comment (zuston) [reply tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java]: done

**Comment by oliverhu** [comment]
@zuston CI tests are failing

**Comment by zuston** [comment]
> @zuston CI tests are failing

sorry, how to see the complete ci test log in circleCI? Now it only show the partial log.

**Comment by oliverhu** [comment]
@zuston  https://app.circleci.com/pipelines/github/linkedin/TonY/342/workflows/fc4d0e23-195f-484d-9bda-94f3a48e20a0/jobs/336 ?

**Comment by zuston** [comment]
> @zuston https://app.circleci.com/pipelines/github/linkedin/TonY/342/workflows/fc4d0e23-195f-484d-9bda-94f3a48e20a0/jobs/336 ?

Yes. So it stuck in **testTonyAMSchedulerShouldPass** method?

**Comment by oliverhu** [comment]
Try run the test locally? Master branch has been green, likely something is broken

**Comment by zuston** [comment]
> Try run the test locally? Master branch has been green, likely something is broken

Sorry, what is **tony.dbloader.instances** in UT testTonyAMSchedulerShouldPass method? @oliverhu

**Comment by oliverhu** [comment]
thanks for the contribution!

**Pull request merged**